File: test/wasm/echo.nim
## Reference `echo` vignette compiled to WASM (T-VIG-ECHO / PAR-01). Mirrors
## src/testing/vignettes.ts EchoVignette: unicasts the bytes back to the sender
## and broadcasts a copy prefixed with the sender id (u16 LE).

import "../../src/vignettes/wasm/vignette"

proc onInit(data: openArray[Byte]) =
  discard data

proc onMessage(senderId: uint32, data: openArray[Byte]): uint32 =
  # Unicast the payload back to the sender.
  emit(uint16(senderId and 0xFFFF'u32), data)

  # Broadcast a copy prefixed with the sender id (u16 LE).
  var tagged = newSeq[Byte](2 + data.len)
  tagged[0] = Byte((senderId shr 0) and 0xFF'u32)
  tagged[1] = Byte((senderId shr 8) and 0xFF'u32)
  for i in 0 ..< data.len:
    tagged[2 + i] = data[i]
  broadcast(tagged)
  0'u32

registerVignetteHandlers(onInit, onMessage)

File: test/wasm/echo.c
/*
 * Reference `echo` vignette in C (T-VIG-ECHO / PAR-01). Mirrors
 * src/testing/vignettes.ts EchoVignette: unicasts the bytes back to the sender,
 * then broadcasts a copy prefixed with the sender id (u16 LE).
 */

#include "wg_vf.h"

#include <stdlib.h>
#include <string.h>

static void on_init(uint8_t *data, uint32_t len) {
  (void)data;
  (void)len;
}

static uint32_t on_message(uint32_t sender_id, uint8_t *data, uint32_t len) {
  wg_vf_emit((uint16_t)(sender_id & 0xFFFF), data, len);

  uint8_t *tagged = (uint8_t *)malloc(len + 2u);
  if (!tagged) {
    return 1; /* allocation failure → sim-fatal */
  }
  tagged[0] = (uint8_t)(sender_id & 0xFF);
  tagged[1] = (uint8_t)((sender_id >> 8) & 0xFF);
  if (len) {
    memcpy(tagged + 2, data, len);
  }
  wg_vf_broadcast(tagged, len + 2u);
  free(tagged);
  return 0;
}

__attribute__((constructor)) static void wg_vf_echo_register(void) {
  wg_vf_handlers h = {0};
  h.on_init = on_init;
  h.on_message = on_message;
  wg_vf_register(&h);
}
